sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/routing/StringPartitionKeyComponent.java

```diff
@@ -94,7 +94,7 @@ public void writeForBinaryEncoding(OutputStream outputStream) {
 
             for (int index = 0; index < (shortString ? this.utf8Value.length : MAX_STRING_BYTES_TO_APPEND + 1); index++) {
                 byte charByte = this.utf8Value[index];
-                if (charByte < 0xFF) charByte++;
```

Comment:
Why do we do this at all?

Reply:
Yea, this was a code that we might have copied from somewhere (may be .Net long time back), until Anu pointed out this issue when we did spot bugs check for v4 during its GA phase. Never really got the chance to fix it, until I started going over the github issues. It fixes this one - https://github.com/Azure/azure-sdk-for-java/issues/9104

Reply:
But why do we add to the original value anyway?

Reply:
Great question, I wish if I had an answer for it :)
May be this history might help - https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB?path=/Product/Microsoft.Azure.Documents/SharedFiles/Routing/StringPartitionKeyComponent.cs&version=GBmaster&line=94&lineEnd=95&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

Dates back to 2016 ;)

Reply:
Here's the whole commit: https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB/commit/71cfa1936b5dbd1f032b2028a6b1461c0401a74c?refName=refs/heads/master&path=/Product/Microsoft.Azure.Documents/SharedFiles/Routing/StringPartitionKeyComponent.cs&_a=compare

Only thing I can guess is that the author wanted to avoid any embedded "0" bytes in the string, since it looks like the binary format being serialized into uses 0 as a terminator.  Therefore it writes "byteVal+1", and the reads and subtracts 1.

Reply:
Also looks like the C# version casts `char` directly to `byte`, and so won't handle any non-ascii strings :-/

Reply:
Interesting, yeah for Java, UTF-8 encoding should handle that for us.